h2o-genmodel/src/main/java/hex/genmodel/algos/ensemble/StackedEnsembleMojoModel.java

```diff
@@ -2,43 +2,90 @@
 
 import hex.genmodel.MojoModel;
 
+import java.util.Arrays;
+
 public class StackedEnsembleMojoModel extends MojoModel {
 
     MojoModel _metaLearner; //Currently only a GLM. May change to be DRF, GBM, XGBoost, or DL in the future
-    MojoModel[] _baseModels; //An array of base models
+    StackedEnsembleMojoSubModel[] _baseModels; //An array of base models
     int _baseModelNum; //Number of base models
+    int _numMappings; // Number of mappings against referential column order
+    double[][] _precalculatedMappings;
 
     public StackedEnsembleMojoModel(String[] columns, String[][] domains, String responseColumn) {
         super(columns, domains, responseColumn);
     }
 
     @Override
     public double[] score0(double[] row, double[] preds) {
+        _precalculatedMappings = new double[_numMappings][];
```

Comment:
I wonder how much time can you save with caching the mappings. In both cases, you are making a copy of the array and you need to touch every element. The difference essentially is whether you iterate sequentially or you use random access. The input rows are typically small (hundreds/thousands of columns) and will fit in a CPU cache.

This is my synthetic benchmark that copies N columns iter-amount of times sequentially and then with random access:

    copyRandom          10       1  avgt   10  ≈ 10⁻⁵           ms/op
    copyRandom          10     100  avgt   10   0.001 ±  0.001  ms/op
    copyRandom        1000       1  avgt   10   0.001 ±  0.001  ms/op
    copyRandom        1000     100  avgt   10   0.056 ±  0.003  ms/op
    copySequential      10       1  avgt   10  ≈ 10⁻⁵           ms/op
    copySequential      10     100  avgt   10   0.001 ±  0.001  ms/op
    copySequential    1000       1  avgt   10  ≈ 10⁻⁴           ms/op
    copySequential    1000     100  avgt   10   0.025 ±  0.002  ms/op

You can see the sequential copy is faster as expected for models with 1000 columns (2x faster). I am not sure if this difference is overall significant compared to an invocation of a GBM/DL/XGB model.

I would probably go with a simpler stateless solution unless there is a significant speed difference. But if you are convinced this idea brings a speedup, please provide a benchmark ;)

Either way - MOJO should ideally be stateless unless there is a strong reason not to.

Reply:
The exact amount of speedup you demonstrate is what I was aiming for. Your benchmark is my benchmark. Didn't think of the statelessness. Reverting the commit right now. The added complexity is not worth it, as mentioned earlier.